src/alpaka/AlpakaCore/getCachingAllocator.h

```diff
@@ -0,0 +1,42 @@
+#ifndef AlpakaCore_getCachingAllocator_h
+#define AlpakaCore_getCachingAllocator_h
+
+#include "AlpakaCore/AllocatorConfig.h"
+#include "AlpakaCore/CachingAllocator.h"
+#include "AlpakaCore/getDevIndex.h"
+
+namespace cms::alpakatools {
+
+  template <typename TDevice>
+  struct CompareDevice {
+    bool operator()(TDevice const& a, TDevice const& b) const { return getDevIndex(a) < getDevIndex(b); }
+  };
+
+  template <typename TDevice, typename TQueue>
+  inline CachingAllocator<TDevice, TQueue>& getCachingAllocator(TDevice const& device) {
+    static std::mutex mutex;
+    static std::map<TDevice, CachingAllocator<TDevice, TQueue>, CompareDevice<TDevice>> allocators;
+
+    auto keyval = allocators.find(device);
```

Comment:
I believe the search needs to be protected with the `mutex` too. (e.g. one thread could be doing `find()` while other is doing `try_emplace()`).

I'm not sure if `std::shared_mutex` would be worth here, or `tbb::concurrent_unordered_map` (which would require defining a hash for the devices?).

Reply:
mhm... maybe just an `std::vector<CachingAllocator>` or `std::unique_ptr<CachingAllocator[]>`, that we fill at the start of the job for all available devices ?
it adds some overhead for the devices that are available but not used - but the allocators don'r allocate any memory by themselves, so not much.

Reply:
> maybe just an `std::vector<CachingAllocator>` or `std::unique_ptr<CachingAllocator[]>`, that we fill at the start of the job for all available devices ?

I agree that would probably be best.

Reply:
The approach used by CUDA for `cudaMallocAsync` _etc._ seems to be that
  - a pool allocator deals with a single device
  - the allocation functions can use an explicit allocator, or fall back to a default per-device allocator
  
Even leaving out the option of using an explicit allocator for the time being, this would suggest that it makes sense to keep the current approach, where the allocator deals with a single device, and there is one allocator per device.